### DISCUSSION TITLE
Best way to parse an Option<T> based on the contents of the field itself

### DISCUSSION THREAD

**Discussion by lfrancke** [sharksforarms/deku#312] [Q&A] [answered]
I have a field like this: `pub retainer_id: Option<i32>` which I'd like to convert to `None` if the parsed value is `-2147483647`. Currently it's parsed as `Some(-2147483647)`

I have a feeling that some combination of `skip`, `cond` and `default` will do the trick but I'm not sure which.

Thank you for this library and your work on it!

**Comment by sharksforarms** [comment, marked as answer]
In this case, [`map`](https://docs.rs/deku/latest/deku/attributes/index.html#map) is how I would approach it since you're still reading the value, but transforming the result based on the value. `skip`, `cond` and `default` can be used to completely bypass reading. Here we always want to read.

```rust
use deku::prelude::*;
use std::convert::{TryFrom, TryInto};

#[derive(Debug, PartialEq, DekuRead, DekuWrite)]
struct DekuTest {
    #[deku(map = "|field: i32| -> Result<_, DekuError> { if field != -2147483647 { Ok(Some(field)) } else { Ok(None) } }")]
    retainer_id: Option<i32>
}

fn main() {
    {
        let test_data = (-2147483647i32).to_le_bytes();

        let test_deku = DekuTest::try_from(test_data.as_slice()).unwrap();

        assert_eq!(
            DekuTest {
                retainer_id: None,
            },
            test_deku
            );

        let test_deku: Vec<u8> = test_deku.try_into().unwrap();
        assert!(test_deku.is_empty());
    }

    {
        let test_data = (-12345i32).to_le_bytes();

        let test_deku = DekuTest::try_from(test_data.as_slice()).unwrap();

        assert_eq!(
            DekuTest {
                retainer_id: Some(-12345),
            },
            test_deku
            );

        let test_deku: Vec<u8> = test_deku.try_into().unwrap();
        assert_eq!(test_data.to_vec(), test_deku);
    }
}
```

**Reply by lfrancke** [reply]
Doh! Thanks again, I did read through the attribute list, I promise. Somehow I didn't make the connection, again.

Thanks for your help!

**Reply by sharksforarms** [reply]
> Doh! Thanks again, I did read through the attribute list, I promise. Somehow I didn't make the connection, again.
> 
> Thanks for your help!

I realize there is a learning curve. I think more examples could help with this. (Actually, maybe these GitHub discussions are a good way of collecting relevant examples!) Thanks for the question!
